### PR TITLE
dataflow: swap Client::recv for Stream

### PR DESCRIPTION
**Pull request by benesch**
The `Client::recv` method is more idiomatically a `Stream`
implementation in Rust async/await. Modeling it as such makes it easier
to compose the `Client` with existing stream combinators.  It also
avoids the pattern in `SelectStream` which incorrectly constructs a new
future on every call to `poll_next`. This defect was spotted by
@petrosagg in #11223; the code happens to work with the current stack of
futures/streams, but might break at any point in the future.

This starts to drive towards the goal laid out in #11078. We don't
actually get separate `Stream` and `Sink` types in this PR, but we at
least get the `Stream` implementation.

Accounting for this refactor in `coordtest` was too difficult to be
worth it, so I opted instead to wholesale remove `coordtest`, as the
comment within `coordtest` offers.

### Motivation

   * This PR refactors existing code.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

**Comment by antiguru** [comment]
LGTM. Dunno if coordtest was useful or not, I'll defer that to someone else.

**Comment by benesch** [comment]
> LGTM. Dunno if coordtest was useful or not, I'll defer that to someone else.

Not to worry! @mjibson has already allowed it:

https://github.com/MaterializeInc/materialize/blob/888300211fc456571cfdaebd69bd71964cee8cd3/src/coordtest/src/lib.rs#L10-L14

File: src/dataflowd/src/lib.rs [hunk 1]
```diff
@@ -145,27 +153,51 @@ impl<S: Client> Client for SplitClient<S> {
...
+impl<C: Client> Stream for SplitClient<C> {
+    type Item = Response;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = &mut *self;
+
+        // Poll {each of the compute clients, the storage client} in turn.
+        for _ in 0..=this.compute_clients.len() {
+            match this.compute_clients.get_mut(&this.recv_cursor) {
+                None => {
```
Comment (frankmcsherry) [inline src/dataflowd/src/lib.rs:167]: These aren't likely to be contiguous, so I think this check is not right. Not trivial to fix afaict. Maybe a `BTreeMap` and picking an offset in the enumeration?

File: src/dataflowd/src/lib.rs [hunk 2]
```diff
@@ -145,27 +153,51 @@ impl<S: Client> Client for SplitClient<S> {
...
+                    this.recv_cursor = 0;
+                    if let Poll::Ready(res) = Pin::new(&mut this.storage_client).poll_next(cx) {
+                        return Poll::Ready(res);
+                    }
+                }
+                Some(compute_client) => {
+                    // We've found a compute client. Set up to poll the next
+                    // compute client on the next turn, then poll the compute
+                    // client.
+                    this.recv_cursor += 1;
```
Comment (frankmcsherry) [inline src/dataflowd/src/lib.rs:180]: I think this is where an opaque `ComputeInstanceId` would have stopped you. Maybe we should do that vs a `typedef`? Idk.

Comment (benesch) [reply src/dataflowd/src/lib.rs:180]: I’m up for that!

File: src/dataflowd/src/lib.rs (hunk 1 above)
Comment (benesch) [reply src/dataflowd/src/lib.rs]: Yeah, `BTreeMap` seems like the way, though oy the implementation is painful. Seems correct, though. Thanks very much for the spot!

**Comment by benesch** [comment]
We're now thinking of this as a perf optimization on top of https://github.com/MaterializeInc/materialize/pull/11317 that we're not interested in unless we really need it!

**Pull request closed without merging**